Doc/library/typing.rst

```diff
@@ -1266,213 +1323,325 @@ These can be used as types in annotations using ``[]``, each having a unique syn
+without the dedicated syntax, as documented below.
 
 .. class:: Generic
 
    Abstract base class for generic types.
 
-   A generic type is typically declared by inheriting from an
-   instantiation of this class with one or more type variables.
-   For example, a generic mapping type might be defined as::
+   A generic type is typically declared by adding a list of type parameters
+   after the class name::
 
-      class Mapping(Generic[KT, VT]):
+      class Mapping[KT, VT]:
           def __getitem__(self, key: KT) -> VT:
               ...
               # Etc.
 
-   This class can then be used as follows::
+   Such a class implicitly inherits from ``Generic``.
+   The runtime semantics of this syntax are discussed in the
+   :ref:`Language Reference <generic-classes>`.
 
-      X = TypeVar('X')
-      Y = TypeVar('Y')
+   This class can then be used as follows::
 
-      def lookup_name(mapping: Mapping[X, Y], key: X, default: Y) -> Y:
+      def lookup_name[X, Y](mapping: Mapping[X, Y], key: X, default: Y) -> Y:
           try:
               return mapping[key]
           except KeyError:
               return default
 
-.. class:: TypeVar
+   Here the brackets after the function name indicate a
+   :ref:`generic function <generic-functions>`.
 
-    Type variable.
+   Generic classes can also be declared by explicitly inheriting from
+   ``Generic``. In this case, the type parameters must be declared
+   separately::
 
-    Usage::
+      KT = TypeVar('KT')
+      VT = TypeVar('VT')
+
+      class Mapping(Generic[KT, VT]):
+          def __getitem__(self, key: KT) -> VT:
+              ...
+              # Etc.
+
+.. class:: TypeVar(name, *constraints, bound=None, covariant=False, contravariant=False, infer_variance=False)
+
+   Type variable.
+
+   Usage::
 
       T = TypeVar('T')  # Can be anything
       S = TypeVar('S', bound=str)  # Can be any subtype of str
       A = TypeVar('A', str, bytes)  # Must be exactly str or bytes
 
-    Type variables exist primarily for the benefit of static type
-    checkers.  They serve as the parameters for generic types as well
-    as for generic function definitions.  See :class:`Generic` for more
-    information on generic types.  Generic functions work as follows::
+   The syntax for :ref:`generic functions <generic-functions>`,
+   :ref:`generic classes <generic-classes>`, and
+   :ref:`generic type aliases <generic-type-aliases>` can be used to
+   create type variables::
+
+      class Sequence[T]:  # T is a TypeVar
+          ...
+
+   This syntax can also be used to create bound and constrained type
+   variables::
 
-       def repeat(x: T, n: int) -> Sequence[T]:
-           """Return a list containing n references to x."""
-           return [x]*n
+      class StrSequence[S: str]:  # S is a TypeVar bound to str
+          ...
+
+
+      class StrOrBytesSequence[A: (str, bytes)]:  # A is a TypeVar constrained to str or bytes
```

Comment:
I think we should do PEP-695 first, old-style manual construction. Something like:

```rst
   The preferred way to construct a type variable is via the dedicated syntax
   for :ref:`generic functions <generic-functions>`,
   :ref:`generic classes <generic-classes>`, and
   :ref:`generic type aliases <generic-type-aliases>`::

      class Sequence[T]:  # T is a TypeVar
          ...

   This syntax can also be used to create bound and constrained type
   variables::

      class StrSequence[S: str]:  # S is a TypeVar bound to str
          ...


      class StrOrBytesSequence[A: (str, bytes)]:  # A is a TypeVar constrained to str or bytes
          ...

   However, if desired, reusable type variables can also be constructed manually, like so::

      T = TypeVar('T')  # Can be anything
      S = TypeVar('S', bound=str)  # Can be any subtype of str
      A = TypeVar('A', str, bytes)  # Must be exactly str or bytes